rgw/sfs: Initial Garbage collector thread.

0xavi0 opened the pull request:
This adds an initial GC approach just starting a thread and running
 a basic remove deleted buckets operation.

The implementation of how the thread is started and stopped are taken from rgw's code for consistency across stores and thinking on the possible merge upstream. 

It uses 2 parameters from rgw:

- `rgw_gc_processor_period` : defines the time GC waits to be triggered (in seconds)
- `rgw_gc_max_objs` : defines the maximum number of items the GC can remove per iteration.

Any possible actions done to the `sqlite` metadata is considered an item (object) when taking into account the maxium number of items to be processed in a single iteration as those actions block the metadata database.

Fixes: https://github.com/aquarist-labs/s3gw/issues/165
Signed-off-by: Xavi Garcia <xavi.garcia@suse.com>


## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

giubacc commented:
I have still to see the implementation; I was thinking: could it make sense to add also a parameter based on a time unit?
A parameter that says that GC can consume only a predetermined amount of time per cycle?
I'm thinking of this because it's not time-deterministic basing the GC cycle on the max number of elements to recycle.
Btw, it's just a food for thought.

0xavi0 commented:
yeah, I think we could also add that.
This is just to add the initial skeleton.
We still need to define the policies we want to follow and everything.

I added the remove deleted buckets so it does something at the beginning, but didn't want to make it more complex on purpose.

giubacc commented:
ok, I started doing some black-box tests without looking at the code.
I'm testing a scenario where:

```
--rgw-gc-processor-period 30
--rgw_gc_max_objs 10
```

1) put 10 buckets
2) put 1 element into every bucket
3) delete all buckets

```
2022-10-20T15:21:19.242+0200 7ffff33126c0 10 rgw UsageLogger: log_usage: TODO
2022-10-20T15:21:19.382+0200 7fffeead46c0  2 garbage collection: garbage collection: start
2022-10-20T15:21:19.382+0200 7fffeead46c0  2 garbage collection: garbage collection: processing with max_objects = 10
2022-10-20T15:21:19.382+0200 7fffeead46c0  2 garbage collection: garbage collection: deleted buckets found = 10
2022-10-20T15:21:19.426+0200 7fffeead46c0  2 garbage collection: garbage collection: stop
2022-10-20T15:21:49.246+0200 7ffff33126c0 10 rgw UsageLogger: log_usage: TODO
2022-10-20T15:21:49.426+0200 7fffeead46c0  2 garbage collection: garbage collection: start
2022-10-20T15:21:49.426+0200 7fffeead46c0  2 garbage collection: garbage collection: processing with max_objects = 10
2022-10-20T15:21:49.426+0200 7fffeead46c0  2 garbage collection: garbage collection: deleted buckets found = 7
2022-10-20T15:21:49.466+0200 7fffeead46c0  2 garbage collection: garbage collection: stop
2022-10-20T15:22:19.246+0200 7ffff33126c0 10 rgw UsageLogger: log_usage: TODO
2022-10-20T15:22:19.466+0200 7fffeead46c0  2 garbage collection: garbage collection: start
2022-10-20T15:22:19.466+0200 7fffeead46c0  2 garbage collection: garbage collection: processing with max_objects = 10
2022-10-20T15:22:19.466+0200 7fffeead46c0  2 garbage collection: garbage collection: deleted buckets found = 4
2022-10-20T15:22:19.502+0200 7fffeead46c0  2 garbage collection: garbage collection: stop
2022-10-20T15:22:49.247+0200 7ffff33126c0 10 rgw UsageLogger: log_usage: TODO
2022-10-20T15:22:49.503+0200 7fffeead46c0  2 garbage collection: garbage collection: start
2022-10-20T15:22:49.503+0200 7fffeead46c0  2 garbage collection: garbage collection: processing with max_objects = 10
2022-10-20T15:22:49.503+0200 7fffeead46c0  2 garbage collection: garbage collection: deleted buckets found = 0
2022-10-20T15:22:49.503+0200 7fffeead46c0  2 garbage collection: garbage collection: stop
2022-10-20T15:23:19.243+0200 7ffff33126c0 10 rgw UsageLogger: log_usage: TODO
```

I think there is something wrong at logging level:

1) it's not true that there were 10+7+4 buckets in the system
2) I'd expect the sum of objects to be 20 not 21

giubacc commented:
oh right, I misinterpreted logs!
Could be maybe useful adding also a log telling how much objects deleted in that iteration?

0xavi0 commented:
> Could be maybe useful adding also a log telling how much objects deleted in that iteration?

Yeah, that could be useful when debugging.
I've added different logs with different log levels so it doesn't flood the output.

Example:
```bash
2022-10-21T10:05:42.582+0200 7f833d25f640  2 garbage collection: start
2022-10-21T10:05:42.582+0200 7f833d25f640 20 garbage collection: garbage collection: processing with max_objects = 10
2022-10-21T10:05:42.582+0200 7f833d25f640 30 garbage collection: deleted buckets found = 14
2022-10-21T10:05:42.582+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket7.1666339021397486531
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (faa192c2-3740-466e-857a-6e7eaebb1148,8)
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: faa192c2-3740-466e-857a-6e7eaebb1148
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket8.1666339022352906509
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (b64e5dbe-6e30-42d9-b894-673e0589b424,9)
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: b64e5dbe-6e30-42d9-b894-673e0589b424
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket9.1666339023111706475
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (859d02c9-c1fc-4667-b8e8-7d8a38c0d518,10)
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: 859d02c9-c1fc-4667-b8e8-7d8a38c0d518
2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket10.1666339023810725076
2022-10-21T10:05:42.586+0200 7f833d25f640  2 garbage collection: stop
```

Detailed logs are showing only on log level 30+...

jecluis reviewed (changes requested):
generally looks good, but a few comments.

giubacc commented on src/rgw/store/sfs/types.cc:
```diff
@@ -86,6 +86,28 @@ void Object::metadata_finish(SFStore *store) {
+void Object::delete_object(SFStore *store) {
+  // delete empty folders going up to the parent path starting at the uuid path
+  auto current_path = store->get_data_path() / path.to_path();
+  while(std::filesystem::is_empty(current_path)) {
```
good point

giubacc replied:
the problem is if the thread storing an object can deschedule after an hypothetical `ensure directory` and if in that case the GC can access to that bucket (is that mutex-like protected?)

0xavi0 replied:
yup, but isn't it nice to have the store fs clean after the GC did its job? :D 
Now seriously, I used that for my tests, but I think we should remove it completely.

I think we can implement the deletion of empty folders in a future policy.
This is going to be a mess because if we block the fs for long the writers are going to be waiting (more delays).

I'm still not sure how to address this, tbh. 

For now it only deletes the object's folder.

giubacc commented on src/rgw/store/sfs/sfs_gc.cc:
```diff
@@ -0,0 +1,173 @@
+    int secs = cct->_conf->rgw_gc_processor_period;
+
+    if (secs <= end.sec())
+      continue; // next round
```
This is why I proposed to add also a param reasoning in terms of time and not in terms of objects.
Btw, If a GC cycle leasts so long that it reach the next triggering I think that you have a configuration problem (aka: please enlarge your period :) )

0xavi0 replied:
yup, I removed that `continue` stuff.
I did a copy from ceph's GC code but I also agree with you this might be too aggresive.

0xavi0 commented on src/rgw/rgw_sal.cc:
```diff
@@ -209,6 +209,10 @@ rgw::sal::Store* StoreManager::init_storage_provider(const DoutPrefixProvider* d
                           << " user in sfs error r=" << r << dendl;
       }
     }
+    if (store->initialize(cct, dpp) < 0) {
```
now `initialize` doesn't do much, but I guess in the future it could. 
anyway, you're right... if we need it in the future we can add it.

changed.

0xavi0 commented on src/rgw/store/sfs/sfs_gc.cc:
```diff
@@ -0,0 +1,173 @@
+  return out << "garbage collection: ";
+}
+
+int SFSGC::process_deleted_buckets(long int &max_objects) {
```
The value WAS in case we added more methods in the future deleting stuff.
What I did is move `max_objects` to be a member and I set the value from rgw config each iteration.
That way I don't need to pass max_objects around.

0xavi0 commented on src/rgw/store/sfs/sfs_gc.cc:
```diff
@@ -0,0 +1,173 @@
+
+    std::unique_lock locker{lock};
+    cond.wait_for(locker, std::chrono::seconds(secs));
+  } while (!gc->going_down());
```
Yeah, it is already checking before the secs calculation...

jecluis replied:
It is indeed. Somehow I missed that bit.

0xavi0 commented on src/rgw/store/sfs/sfs_gc.cc:
```diff
@@ -0,0 +1,173 @@
+void *SFSGC::GCWorker::entry() {
+  do {
+    utime_t start = ceph_clock_now();
+    ldpp_dout(dpp, 2) << "start" << dendl;
```
changed

giubacc commented:
> > Could be maybe useful adding also a log telling how much objects deleted in that iteration?
> 
> Yeah, that could be useful when debugging. I've added different logs with different log levels so it doesn't flood the output.
> 
> Example:
> 
> ```shell
> 2022-10-21T10:05:42.582+0200 7f833d25f640  2 garbage collection: start
> 2022-10-21T10:05:42.582+0200 7f833d25f640 20 garbage collection: garbage collection: processing with max_objects = 10
> 2022-10-21T10:05:42.582+0200 7f833d25f640 30 garbage collection: deleted buckets found = 14
> 2022-10-21T10:05:42.582+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket7.1666339021397486531
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (faa192c2-3740-466e-857a-6e7eaebb1148,8)
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: faa192c2-3740-466e-857a-6e7eaebb1148
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket8.1666339022352906509
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (b64e5dbe-6e30-42d9-b894-673e0589b424,9)
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: b64e5dbe-6e30-42d9-b894-673e0589b424
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket9.1666339023111706475
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting version: (859d02c9-c1fc-4667-b8e8-7d8a38c0d518,10)
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting object: 859d02c9-c1fc-4667-b8e8-7d8a38c0d518
> 2022-10-21T10:05:42.586+0200 7f833d25f640 30 garbage collection: Deleting bucket: bucket10.1666339023810725076
> 2022-10-21T10:05:42.586+0200 7f833d25f640  2 garbage collection: stop
> ```
> 
> Detailed logs are showing only on log level 30+...

good, but IMO these levels are too much high: `garbage collection: deleted buckets found` shoud be 10 or something similar.

giubacc commented:
Probably this is not dangerous, but I noticed that the first cycle of GC happens just after the process starts and interweaves with the init thread:

```
2022-10-21T16:00:51.981+0200 7ffff3b39e40 10 rgw main: initialize
2022-10-21T16:00:51.981+0200 7fffeead36c0  2 garbage collection: > sfs garbage collector worker::entry start
2022-10-21T16:00:51.981+0200 7fffeead36c0 20 garbage collection: garbage collection: processing with max_objects = 10
2022-10-21T16:00:51.981+0200 7ffff3b39e40 20 RGW hostnames: 
2022-10-21T16:00:51.981+0200 7ffff3b39e40 20 RGW S3website hostnames: 
2022-10-21T16:00:51.981+0200 7ffff3b39e40 10 get_sync_module: TODO
2022-10-21T16:00:51.985+0200 7ffff3b39e40 10 get_sync_module: TODO
2022-10-21T16:00:51.985+0200 7fffeead36c0  2 garbage collection: > sfs garbage collector worker::entry stop
```

I repeat, I don't think this is dangerous, but is this intended?

0xavi0 commented:
log levels changed to 10
```bash
2022-10-21T16:11:25.107+0200 7f12c423d640  2 garbage collection: > GCWorker::entry start
2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process garbage collection: processing with max_objects = 50
2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process_deleted_buckets deleted buckets found = 10
2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9f1759be-ebdf-4554-bf7a-6948c5be4328,1)
2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9f1759be-ebdf-4554-bf7a-6948c5be4328
2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket1.1666361458691050416
2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a5c6e49d-4b45-498c-aaf3-596b390f4f3b,2)
2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a5c6e49d-4b45-498c-aaf3-596b390f4f3b
2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket2.1666361459788390592
2022-10-21T16:11:25.123+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e8687e9a-0e25-42c9-9476-275d68a01af5,3)
2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e8687e9a-0e25-42c9-9476-275d68a01af5
2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket3.1666361460907305882
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9c513ba3-b90e-4483-a247-def4dc603459,4)
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9c513ba3-b90e-4483-a247-def4dc603459
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket4.166636146257766139
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (04b862c0-93cc-46fc-94f3-99f197a4fb07,5)
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 04b862c0-93cc-46fc-94f3-99f197a4fb07
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket5.166636146371594851
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (95896a16-a049-4122-a7c3-4f0875f7eb3b,6)
2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 95896a16-a049-4122-a7c3-4f0875f7eb3b
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket6.1666361464306001114
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e22dcf26-c642-4826-ae50-f0833ec75b4c,7)
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e22dcf26-c642-4826-ae50-f0833ec75b4c
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket7.1666361465264412860
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (6a31f826-e486-4591-9cca-e47b426f7251,8)
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 6a31f826-e486-4591-9cca-e47b426f7251
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket8.1666361466453302512
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a55f1298-0f67-41c7-aa27-b2cee04bdaee,9)
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a55f1298-0f67-41c7-aa27-b2cee04bdaee
2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket9.1666361467401556259
2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (bb2b8620-c27c-436a-9fd4-078d9fb63901,10)
2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: bb2b8620-c27c-436a-9fd4-078d9fb63901
2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket10.1666361468472669007
2022-10-21T16:11:25.139+0200 7f12c423d640  2 garbage collection: > GCWorker::entry stop
```

0xavi0 commented:
> Probably this is not dangerous, but I noticed that the first cycle of GC happens just after the process starts and interweaves with the init thread:
> 
> ```
> 2022-10-21T16:00:51.981+0200 7ffff3b39e40 10 rgw main: initialize
> 2022-10-21T16:00:51.981+0200 7fffeead36c0  2 garbage collection: > sfs garbage collector worker::entry start
> 2022-10-21T16:00:51.981+0200 7fffeead36c0 20 garbage collection: garbage collection: processing with max_objects = 10
> 2022-10-21T16:00:51.981+0200 7ffff3b39e40 20 RGW hostnames: 
> 2022-10-21T16:00:51.981+0200 7ffff3b39e40 20 RGW S3website hostnames: 
> 2022-10-21T16:00:51.981+0200 7ffff3b39e40 10 get_sync_module: TODO
> 2022-10-21T16:00:51.985+0200 7ffff3b39e40 10 get_sync_module: TODO
> 2022-10-21T16:00:51.985+0200 7fffeead36c0  2 garbage collection: > sfs garbage collector worker::entry stop
> ```
> 
> I repeat, I don't think this is dangerous, but is this intended?

I guess it's useful when restarting the gw... if there is stuff ready to be deleted  by the GC.
We could wait for the N seconds for the first iteration or trigger it at the beginning.
That code is from `ceph` and looks like they decided to do it when it starts.

I don't think that's bad. When the GW starts is a good moment for cleaning stuff. (Normally less user activity)

jecluis commented:
> log levels changed to 10
> 
> ```shell
> 2022-10-21T16:11:25.107+0200 7f12c423d640  2 garbage collection: > GCWorker::entry start
> 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process garbage collection: processing with max_objects = 50
> 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process_deleted_buckets deleted buckets found = 10
> 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9f1759be-ebdf-4554-bf7a-6948c5be4328,1)
> 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9f1759be-ebdf-4554-bf7a-6948c5be4328
> 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket1.1666361458691050416
> 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a5c6e49d-4b45-498c-aaf3-596b390f4f3b,2)
> 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a5c6e49d-4b45-498c-aaf3-596b390f4f3b
> 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket2.1666361459788390592
> 2022-10-21T16:11:25.123+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e8687e9a-0e25-42c9-9476-275d68a01af5,3)
> 2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e8687e9a-0e25-42c9-9476-275d68a01af5
> 2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket3.1666361460907305882
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9c513ba3-b90e-4483-a247-def4dc603459,4)
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9c513ba3-b90e-4483-a247-def4dc603459
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket4.166636146257766139
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (04b862c0-93cc-46fc-94f3-99f197a4fb07,5)
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 04b862c0-93cc-46fc-94f3-99f197a4fb07
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket5.166636146371594851
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (95896a16-a049-4122-a7c3-4f0875f7eb3b,6)
> 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 95896a16-a049-4122-a7c3-4f0875f7eb3b
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket6.1666361464306001114
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e22dcf26-c642-4826-ae50-f0833ec75b4c,7)
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e22dcf26-c642-4826-ae50-f0833ec75b4c
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket7.1666361465264412860
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (6a31f826-e486-4591-9cca-e47b426f7251,8)
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 6a31f826-e486-4591-9cca-e47b426f7251
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket8.1666361466453302512
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a55f1298-0f67-41c7-aa27-b2cee04bdaee,9)
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a55f1298-0f67-41c7-aa27-b2cee04bdaee
> 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket9.1666361467401556259
> 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (bb2b8620-c27c-436a-9fd4-078d9fb63901,10)
> 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: bb2b8620-c27c-436a-9fd4-078d9fb63901
> 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket10.1666361468472669007
> 2022-10-21T16:11:25.139+0200 7f12c423d640  2 garbage collection: > GCWorker::entry stop
> ```

Okay, I hate to put you through a "do this", "do that" kind of thing, but I think having those messages at level 10 is waaay too low. If there are a lot of objects to be GC'ed, this is going to spam the logs for no reason. I'd just have a "deleting bucket 'foo' with 1234 objects" message at start and another acknowledging each finished cycle, and have the individual delete messages at 30 like you had before.

giubacc commented on src/rgw/store/sfs/sfs_gc.cc:
```diff
@@ -0,0 +1,193 @@
+    secs -= end.sec();
+
+    std::unique_lock locker{lock};
+    cond.wait_for(locker, std::chrono::seconds(secs));
```
is this `wait_for()` for filling the remaining amount of time not consumed for the cycle, correct?
what if `secs` is negative because, for any reason: `secs - elapsed` is a negative number?

0xavi0 replied:
It doesn't wait (tested it)

While I understand the concerns around this... We're talking about seconds precision here.

IMO, the GC should not trigger too often (in ceph defaults to 1h).

And even if we make it trigger every second.... (which I think is not a realistic scenario) if it takes more that 1 second to cleanup... we still have the `rgw_gc_max_objs` parameter to limit the GC... (and possibly another future policy based on elapsed time) and if it's still taking more than 1 second I would be worried about the whole GC, tbh. 

Let's say we configure it to run every 5 mins...
If the GC takes 5 mins to do its job... I would better make the GC delete itself for being completely incompetent. :) or maybe change the `rgw_gc_max_objs` for being crazy high... (or maybe we have thousands of users uploading thousands of objects and buckets and starving the GC thread)

Having said that. I added some code in case the GC is so slow.... to make it wait for 1 period. :)

giubacc commented on src/test/rgw/sfs/test_rgw_sfs_gc.cc:
```diff
@@ -0,0 +1,316 @@
+  // the object is finally removed
+  EXPECT_EQ(0, getNumberObjectsForBucket("test_bucket_2", store->db_conn));
+  EXPECT_EQ(1, getNumberObjectsForBucket("test_bucket_1", store->db_conn));
+}
```
let's remember to add this suite to the CI

0xavi0 replied:
Sure, I think we should move the tests as soon as possible to the ceph repo.

giubacc commented:
> > log levels changed to 10
> > ```shell
> > 2022-10-21T16:11:25.107+0200 7f12c423d640  2 garbage collection: > GCWorker::entry start
> > 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process garbage collection: processing with max_objects = 50
> > 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::process_deleted_buckets deleted buckets found = 10
> > 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9f1759be-ebdf-4554-bf7a-6948c5be4328,1)
> > 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9f1759be-ebdf-4554-bf7a-6948c5be4328
> > 2022-10-21T16:11:25.107+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket1.1666361458691050416
> > 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a5c6e49d-4b45-498c-aaf3-596b390f4f3b,2)
> > 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a5c6e49d-4b45-498c-aaf3-596b390f4f3b
> > 2022-10-21T16:11:25.111+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket2.1666361459788390592
> > 2022-10-21T16:11:25.123+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e8687e9a-0e25-42c9-9476-275d68a01af5,3)
> > 2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e8687e9a-0e25-42c9-9476-275d68a01af5
> > 2022-10-21T16:11:25.127+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket3.1666361460907305882
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (9c513ba3-b90e-4483-a247-def4dc603459,4)
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 9c513ba3-b90e-4483-a247-def4dc603459
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket4.166636146257766139
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (04b862c0-93cc-46fc-94f3-99f197a4fb07,5)
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 04b862c0-93cc-46fc-94f3-99f197a4fb07
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket5.166636146371594851
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (95896a16-a049-4122-a7c3-4f0875f7eb3b,6)
> > 2022-10-21T16:11:25.131+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 95896a16-a049-4122-a7c3-4f0875f7eb3b
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket6.1666361464306001114
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (e22dcf26-c642-4826-ae50-f0833ec75b4c,7)
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: e22dcf26-c642-4826-ae50-f0833ec75b4c
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket7.1666361465264412860
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (6a31f826-e486-4591-9cca-e47b426f7251,8)
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: 6a31f826-e486-4591-9cca-e47b426f7251
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket8.1666361466453302512
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (a55f1298-0f67-41c7-aa27-b2cee04bdaee,9)
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: a55f1298-0f67-41c7-aa27-b2cee04bdaee
> > 2022-10-21T16:11:25.135+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket9.1666361467401556259
> > 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_versioned_object Deleting version: (bb2b8620-c27c-436a-9fd4-078d9fb63901,10)
> > 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_object Deleting object: bb2b8620-c27c-436a-9fd4-078d9fb63901
> > 2022-10-21T16:11:25.139+0200 7f12c423d640 10 garbage collection: > SFSGC::delete_bucket Deleting bucket: bucket10.1666361468472669007
> > 2022-10-21T16:11:25.139+0200 7f12c423d640  2 garbage collection: > GCWorker::entry stop
> > ```
> 
> Okay, I hate to put you through a "do this", "do that" kind of thing, but I think having those messages at level 10 is waaay too low. If there are a lot of objects to be GC'ed, this is going to spam the logs for no reason. I'd just have a "deleting bucket 'foo' with 1234 objects" message at start and another acknowledging each finished cycle, and have the individual delete messages at 30 like you had before.

IMO, log `process_deleted_buckets deleted buckets found` has a dignity to assume a `15` log level. Agree on other detail logs to assume `>=20` value.

jecluis commented:
> IMO, log `process_deleted_buckets deleted buckets found` has a dignity to assume a `15` log level. Agree on other detail logs to assume `>=20` value.

Lets not do halfsies on log levels please. Ceph's log levels are confusing and ad hoc enough as they are. No one is going to remember to set log level 15 to debug this should they need it in production, unless they have insight into the code. People usually do multiples of 10, unless there's documentation for things, or have already developed tribal knowledge about what they are debugging.

Either do level 10 or level 20. Not level 15.

giubacc commented:
> > IMO, log `process_deleted_buckets deleted buckets found` has a dignity to assume a `15` log level. Agree on other detail logs to assume `>=20` value.
> 
> Lets not do halfsies on log levels please. Ceph's log levels are confusing and ad hoc enough as they are. No one is going to remember to set log level 15 to debug this should they need it in production, unless they have insight into the code. People usually do multiples of 10, unless there's documentation for things, or have already developed tribal knowledge about what they are debugging.
> 
> Either do level 10 or level 20. Not level 15.

There are already usage cases for 5 and 15 in the codebase, is there an explicit rule in Ceph community to use 10 multiples? 
Ok if we want to define a rule downstream for sfs.

jecluis commented:
> There are already usage cases for 5 and 15 in the codebase, is there an explicit rule in Ceph community to use 10 multiples? Ok if we want to define a rule downstream for sfs.

No, there are no explicit guidelines, and that is the main problem. Usually, rule of thumb, the more verbose messages are/the more annoying, the higher the debug level should be. In the monitors we have a lot of verbose messages being dumped at 30, or 20, depending on how often they happen or are needed.

And yes, there is plenty at 5, and 15, and then something between them. And at 1, and at 2. And on some components you barely have anything at 10, but at 1, 2, 5, etc; on others you usually have 0, 1, 5, 10, 20, etc. It really depends. Hence why I'm suggesting not to do that and do the obvious instead. When debugging issues, people do use to go from 10 to 20, unless a developer tells them otherwise. I've never seen a customer issue starting with someone saying "let's set debug levels at 5", but usually goes with "let's set debug levels at 10", and maybe at 20. Because those are easy to remember. 5, 15, 25, whatever, aren't.

giubacc commented:
> > There are already usage cases for 5 and 15 in the codebase, is there an explicit rule in Ceph community to use 10 multiples? Ok if we want to define a rule downstream for sfs.
> 
> No, there are no explicit guidelines, and that is the main problem. Usually, rule of thumb, the more verbose messages are/the more annoying, the higher the debug level should be. In the monitors we have a lot of verbose messages being dumped at 30, or 20, depending on how often they happen or are needed.
> 
> And yes, there is plenty at 5, and 15, and then something between them. And at 1, and at 2. And on some components you barely have anything at 10, but at 1, 2, 5, etc; on others you usually have 0, 1, 5, 10, 20, etc. It really depends. Hence why I'm suggesting not to do that and do the obvious instead. When debugging issues, people do use to go from 10 to 20, unless a developer tells them otherwise. I've never seen a customer issue starting with someone saying "let's set debug levels at 5", but usually goes with "let's set debug levels at 10", and maybe at 20. Because those are easy to remember. 5, 15, 25, whatever, aren't.

As far as I'm concerned, I haven't a special "problem" using a number instead of another and personally I like very much 5 multiplies, but this is just a personal case.

giubacc commented:
apart from the needed rebase, I'd say this can be considered good for merging.
